components/raftstore/src/store/worker/pd.rs

```diff
@@ -1084,6 +1169,247 @@ fn send_destroy_peer_message(
     }
 }
 
+const DEFAULT_QPS_THRESHOLD: u32 = 500;
+const DETECT_TIMES: u32 = 10;
+pub const TOP_N: usize = 10;
+const DETECT_INTERVAL: Duration = Duration::from_secs(1);
+const MIN_SAMPLE_NUM: i32 = 100;
+const DEFAULT_SPLIT_SCORE: f64 = 0.25;
+
+pub struct Sample {
+    pub key: Vec<u8>,
+    pub left: i32,
+    pub contained: i32,
+    pub right: i32,
+}
+
+impl Sample {
+    fn new(key: &[u8]) -> Sample {
+        Sample {
+            key: key.to_owned(),
+            left: 0,
+            contained: 0,
+            right: 0,
+        }
+    }
+}
+
+pub struct KeyRange {
+    pub start_key: Vec<u8>,
+    pub end_key: Vec<u8>,
+    pub qps: u64,
+}
+
+impl KeyRange {
+    fn new(start_key: &[u8], end_key: &[u8]) -> KeyRange {
+        KeyRange {
+            start_key: start_key.to_owned(),
+            end_key: end_key.to_owned(),
+            qps: 0,
+        }
+    }
+
+    fn contains(&self, key: &[u8]) -> bool {
+        key.cmp(&self.start_key) == Ordering::Greater && key.cmp(&self.end_key) == Ordering::Less
```

Comment:
If key==start_key==end_key, it returns false.

Reply:
Thanks